controllers/authorController.js

```diff
@@ -15,35 +19,196 @@ exports.author_list = function(req, res) {
+  
+  // Process request after validation and sanitization.
+  (req, res, next) => {
+
+      //provide edited dates
+      
+      let author_temp = {
+        first_name: req.body.first_name,
+        family_name: req.body.family_name,
+        date_birth_form: DateTime.fromJSDate(req.body.date_of_birth).toISODate(),
+        date_death_form: DateTime.fromJSDate(req.body.date_of_death).toISODate()
+      }   
+      
+      // Extract the validation errors from a request.
+      const errors = validationResult(req);
+      
+
+      if (!errors.isEmpty()) {
+          // There are errors. Render form again with sanitized values/errors messages.
+          res.render('author_form', { title: 'Create Author', author: author_temp, errors: errors.array() });
+          return;
+      }
+      else {
+          // Data from form is valid.
+
+          // Create an Author object with escaped and trimmed data.
+          let author = new Author(
+              {
+                  first_name: req.body.first_name,
+                  family_name: req.body.family_name,
+                  date_of_birth: req.body.date_of_birth,
+                  date_of_death: req.body.date_of_death
+              });
+          author.save(function (err) {
+              if (err) { return next(err); }
+              // Successful - redirect to new author record.
+              res.redirect(author.url);
+          });
+      }
+  }
+];
 
 // Display Author delete form on GET.
-exports.author_delete_get = function(req, res) {
-    res.send('NOT IMPLEMENTED: Author delete GET');
+exports.author_delete_get = function(req, res, next) {
+
+  async.parallel({
+      author: function(callback) {
+          Author.findById(req.params.id).exec(callback)
+      },
+      authors_blogposts: function(callback) {
+        BlogPost.find({ 'author': req.params.id }).exec(callback)
+      },
+  }, function(err, results) {
+      if (err) { return next(err); }
+      if (results.author==null) { // No results.
+          res.redirect('/blog/authors');
+      }
+      // Successful, so render.
+      res.render('author_delete', { title: 'Delete Author', author: results.author, author_blogposts: results.authors_blogposts } );
+  });
+
 };
 
 // Handle Author delete on POST.
-exports.author_delete_post = function(req, res) {
-    res.send('NOT IMPLEMENTED: Author delete POST');
+exports.author_delete_post = function(req, res, next) {
+  async.parallel({
+      author: function(callback) {
+        Author.findById(req.body.authorid).exec(callback)
+      },
+      authors_blogposts: function(callback) {
+        BlogPost.find({ 'author': req.body.authorid }).exec(callback)
+      },
+  }, function(err, results) {
+      if (err) { return next(err); }
+      
+      // Success
+      if (results.authors_blogposts.length > 0) {
+          // Author has books. Render in same way as for GET route.
```

Comment:
S/books/posts/

Consider adding some documentation here to make it clear that before an author can be deleted, all of their blog posts need to be deleted.

Reply:
Added documentation on authorController.js.  Need to review other controllers and add documentation.

Reply:
Reviewed other controllers and added documentation.